src/main/java/fi/aalto/cs/apluscourses/intellij/model/IntelliJModelFactory.java

```diff
@@ -22,11 +28,43 @@ public IntelliJModelFactory(@NotNull Project project) {
   public Course createCourse(@NotNull String name,
                              @NotNull List<Module> modules,
                              @NotNull Map<String, String> requiredPlugins) {
-    return new IntelliJCourse(name, modules, requiredPlugins, project);
+    IntelliJCourse course = new IntelliJCourse(name, modules, requiredPlugins, project);
+    // Add a module change listener with the created course instance to the project
+    project.getMessageBus().connect().subscribe(ProjectTopics.MODULES, new ModuleListener() {
+      @Override
+      public void moduleRemoved(@NotNull Project project,
+                                @NotNull com.intellij.openapi.module.Module projectModule) {
+        Module module = course.getModuleOpt(projectModule.getName());
+        if (module != null) {
+          course.onModuleRemove(module);
+        }
+      }
+    });
+    project.getMessageBus().connect().subscribe(VirtualFileManager.VFS_CHANGES,
+        new BulkFileListener() {
+          @Override
+          public void after(@NotNull List<? extends VFileEvent> events) {
+            for (VFileEvent event : events) {
+              if (event instanceof VFileDeleteEvent) {
+                String deletedFileName = event.getFile().getName();
```

Comment:
Method invocation `getName()` may produce NPE

Reply:
You're right, I'll add a check